### ISSUE TITLE
Feature Request: Enable native applications writing to PS native streams (Info, Warning, etc.)

### ISSUE THREAD

**Issue by bgshacklett** [PowerShell/PowerShell#15564]
## Summary of the new feature/enhancement

As a software developer, I want to be able to write to the native PowerShell streams so that I can ensure that output is treated appropriately and provide a more native experience to my users.

I've seen, and encountered, a number of issues reported across multiple repositories (and my own use cases) where people are dealing with the fact that writing to `stderr` from a native application may result in an error. This has led to many suggestions for workarounds and changes to PowerShell's behavior which may not fit with the original goals and design.

Rather than changing the behavior of how PowerShell handles Stderr (or perhaps in addition to), it would be quite useful for developers using Python, Go, Haskell, et al., to be able to write to additional streams, such as Warning, Info, Verbose, etc.. This would facilitate development of native applications which feel closer in behavior to PowerShell Cmdlets. I expect this may also remove some barriers to adoption, which exist amongst POSIX shell purists. 

## Proposed technical implementation details (optional)

I expect that this enhancement would take the form of multiple phases:
1. Ground work to enable the behavior and a proof of concept, of some sort, showing a native application writing to, for example, the Informational stream.
2. A documented recommendation for developers of languages which result in native executables, which explains how to write libraries which enable output to PS streams.

Additional goals:
3. A documented standard for the "protocol", or mechanism, which enables native streams. This might enable the same architecture to be implemented in other shells.
4. Reference libraries for languages which may be strategic to Microsoft, e.g.: Python.
5. Extend the solution to enable writing object streams from native applications, rather than pure text.

**Comment by jborean93** [comment]
This is actually possible today but there are a few caveats. Your native app can output to stdout in CLIXML the objects you want to return back to PowerShell. Here is a basic example of that when using bash (any app will do though)

```powershell
Set-Content -Path /tmp/test.sh -Value @'
#!/usr/bin/env bash

echo '#< CLIXML'
echo -n '<Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04">'
echo -n  '<S S="Output">output string</S>'
echo -n  '<Obj S="Output" RefId="0"><TN RefId="0"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><S N="Foo">Bar</S></MS></Obj>'
echo -n  '<S S="Verbose">verbose msg</S>'
echo -n '</Objs>'
'@

$out = bash /tmp/test.sh
```

When running that I get the following

![image](https://user-images.githubusercontent.com/8462645/121739931-ac5baf00-cb3f-11eb-93c4-dfa254b88399.png)


What's happening is that pwsh sees the start of the output is `# <CLIXML` and then parses the rest of the output as actual objects. It allows you to write to specific streams as well as write complex objects back which is pretty nice. The downsides of this

* The native app needs to be aware of this - `pwsh` has the `-OutputFormat XML` that does the same thing, a native app could do the same
* You explicitly need to capture/redirect the output when calling the native app
  * Not doing so means the output is written directly to the console so the end user will see the CLIXML
  * Capturing is essentially doing something like `$var = native` or pipelining/redirection in pwsh
* CLIXML is very verbose
  * Simple strings are fine but dealing with complex objects gets messy very quickly

**Comment by 237dmitry** [comment]
> As far as my own understanding goes, native applications on Windows, MacOS, and Linux generally only have access to two streams: stdout, and stderror.

And `stdin`. And `3..8` descriptors may be used for arbitrary output (for bash):
```bash
exec 3>file
echo "test" >&3
cat file
```

**Comment by iSazonov** [comment]
We discussed such proposal previously and rejected because native application sometimes use error output handle not for writing errors but for other purposes.

**Comment by jborean93** [comment]
I’m not sure what else needs to be done, it’s entirely possible as evident by my previous comment.

**Comment by rkeithhill** [comment]
> I've seen, and encountered, a number of issues reported across multiple repositories (and my own use cases) where people are dealing with the fact that writing to stderr from a native application may result in an error. 

This issue is being addressed via a feature which is experiment in the 7.2 previews `PSNotApplyErrorActionToStderr`.

**Comment by bgshacklett** [comment]
> We discussed such proposal previously and rejected because native application sometimes use error output handle not for writing errors but for other purposes.

That's one thing that led me to open this request. I think the preferred solution, for those who wish to target PS, would be to make use of the Info stream, but at present it doesn't seem feasible. If a mechanism for that could be surfaced, such applications could handle streams appropriately per shell, and we might even see other shells add additional streams some day.

@jborean93 Your suggestion is really interesting, and I'll probably do some exploration on that front. One of the caveats, however, makes it quite difficult to rely on the current mechanism:
> You explicitly need to capture/redirect the output when calling the native app
>
>    Not doing so means the output is written directly to the console so the end user will see the CLIXML
>    Capturing is essentially doing something like $var = native or pipelining/redirection in pwsh

While possible, I believe it would be deemed infeasible, in most scenarios, to assume that output would be captured in this way. This is going to be outside of the control of most developers, unless they decide to distribute a shim with their application. While that might be technically feasible, imposing that on the majority of those who are writing native applications seems unlikely to bring about any useful results.

Regarding the other two points:
> *    The native app needs to be aware of this - pwsh has the -OutputFormat XML that does the same thing, a native app >could do the same
> *    CLIXML is very verbose
>        Simple strings are fine but dealing with complex objects gets messy very quickly

While not perfect, I think these are solve-able, with the first being the most difficult (because environment detection is almost always painful). Dealing with CLIXML wouldn't be too bad if libraries were available in developers' chosen languages.

**Comment by jborean93** [comment]
> While possible, I believe it would be deemed infeasible, in most scenarios, to assume that output would be captured in this way.

Agreed, luckily at least on Windows it is definitely possible to detect whether your app has redirected output (pwsh is capturing it) or it's connected directly to the console. This allows you to have your native app determine whether to output as per usual or whether it should output in CLIXML. I would recommend you still put it behind a certain argument/switch when calling the native app just to be explicit just like pwsh has `-OutputFormat XML`.

To detect whether your `stdout` is being captured you need to call [GetStdHandle](https://docs.microsoft.com/en-us/windows/console/getstdhandle) for `-11` (`stdout`) then call [GetFileType](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfiletype). If the value is `FILE_TYPE_CHAR` you know no redirection has occurred and if it's `FILE_TYPE_PIPE` you know it has been redirected and pwsh is capturing it.

This somewhat makes sense as you don't want to be writing CLIXML to the stdout yourself for any actual output as it will be written straight to the console. It does suck if all you want to do is write to the verbose/warning/debug/info stream but the workaround is as simple as doing something like `native.exe | Out-String`. Considering you will probably have this behind some flag like `native.exe -OutFormat Pwsh` it's probably not too hard of an ask to make sure you are piping it into pwsh.

It's unfortunate but simply something you just need to adjust when calling your native app.

**Comment by vexx32** [comment]
I think we'd need to sort out a fix for https://github.com/PowerShell/PowerShell/issues/15363 before we can really call this "solved"; currently CLIXML isn't always deserialised by PS when received, and I think that's pretty undesirable. It's very non-obvious to the user that simply capturing the output will magically deserialise it properly.

Another thing was mentioned here that is worth reinforcing as well -- something that would make this significantly easier is creating a standalone CLIXML library for other folks to use when structuring their output, and provide abstractions for writing to the distinct streams.

Even if only a C# library is provided initially, if it's open source others can take pointers from there and copy the implementation into another language to allow this to work for any native application with minimal friction.

**Comment by jborean93** [comment]
> I think we'd need to sort out a fix for #15363 before we can really call this "solved"

Somewhat agreed, the trouble is that to solve #15363 would mean that powershell always processes the `stdout` leading to the existing problems with how it encodes the bytes from this stream. There are many issues with this and there's no silver bullet to solve them all so I personally doubt this is something that is really feasible. It also doesn't really make much sense to convert an object from CLIXML when all it is going to do is then output it to the host. The only major disadvantage I can see from the optional CLIXML parsing is that you don't write data to the various streams. Even then this isn't too big of a problem IMO because

* Streams just typically write to the host in the default session, the native app could do the same thing in this case
* If you are running it in a PowerShell pipeline and are inspecting the streams manually then the output is captured by pwsh anyway so the CLIXML is parsed

It's not ideal but it's possible today for a native app to determine whether to print CLIXML or just text. I would have to double check on Linux around this but I don't see any reasons why it wouldn't be possible

**Comment by SeeminglyScience** [comment]
If a native application is *targeting* PowerShell then I'd expect a binary module instead.  If they aren't explicitly targeting PowerShell, then any solution is going to require quite a bit of extra complexity and overhead.  Either output will need to be serialized in a way PowerShell recognizes (which would then require PowerShell to analyze the output of every native app) or some sort of IPC would have to be opened between them.

To me it seems like the ideal use case for the proposed feature is something the dev can tack onto their existing application to make it PowerShell friendly rather than being required in run in PowerShell.  I have a hard time imagining how this feature could be implemented in a way that would make the extra complexity worth while.

**Comment by antaln** [comment]
> This is actually possible today but there are a few caveats. Your native app can output to stdout in CLIXML the objects you want to return back to PowerShell Here is a basic example of that when using bash (any app will do though)

@jborean93 , brilliant! Any way to output a progress record in this way?

**Comment by bgshacklett** [comment]
Is there a spec for CLIXML anywhere?
